eventsourced/src/lib.rs

```diff
@@ -66,204 +55,111 @@ use tokio::{
 };
 use tracing::{debug, error, instrument};
 
-type BoxedAny = Box<dyn Any + Send>;
-type BoxedHandleCmd<E> = Box<
-    dyn Fn(&BoxedAny, &<E as EventSourced>::Id, &E) -> Result<<E as EventSourced>::Evt, BoxedAny>
-        + Send,
->;
-type BoxedReply<E> = Box<dyn Fn(BoxedAny, &E) -> BoxedAny + Send + Sync>;
-type BoxedCmdFns<E> = HashMap<TypeId, (BoxedHandleCmd<E>, BoxedReply<E>)>;
+type Channel<E> = mpsc::Sender<(
+    <E as EventSourced>::Cmd,
+    oneshot::Sender<Result<<E as EventSourced>::Reply, <E as EventSourced>::Error>>,
+)>;
 
-/// The state of an event sourced entity as well as its event handling (which transforms the state).
+/// Command and event handling for an event sourced entity.
 pub trait EventSourced {
-    /// The Id type.
-    type Id: Debug + Clone + Send;
-
-    /// The event type.
-    type Evt: Debug + Send + Sync + 'static;
-
-    /// The type name.
-    const TYPE_NAME: &'static str;
-
-    /// The event handler.
-    fn handle_evt(self, evt: Self::Evt) -> Self;
-}
+    /// Id type.
+    type Id: Debug + Send + 'static;
 
-/// A command for the given [EventSourced] implementation, defining its handling and replying.
-pub trait Cmd<E>
-where
-    Self: 'static,
-    E: EventSourced,
-{
-    /// The type for rejecting this command.
-    type Error: Send + 'static;
+    /// Command type.
+    type Cmd: Debug + Send + Sync + 'static;
 
-    /// The type for replies.
-    type Reply: Send + 'static;
+    /// Event type.
+    type Evt: Debug + Send + Sync;
 
-    /// The command handler, taking this command, and references to the ID and the state of
-    /// the event sourced entity, either rejecting this command via [Self::Error] or returning an
-    /// event.
-    fn handle_cmd(&self, id: &E::Id, state: &E) -> Result<E::Evt, Self::Error>;
+    /// State type.
+    type State: Debug + Default + Send + Sync + 'static;
 
-    /// The reply function, which is applied if the command handler has returned an event (as
-    /// opposed to a rejection) and after that has been persisted successfully.
-    fn reply(self, state: &E) -> Self::Reply;
-}
+    /// Type for partial reply data.
+    type ReplyInfo: Send;
 
-/// A handle representing a spawned [EventSourced] entity, which can be used to pass it commands
-/// which have been registerd before via [EventSourcedEntity::cmd].
-#[derive(Debug, Clone)]
-pub struct EntityRef<E, T>
-where
-    E: EventSourced,
-{
-    cmd_in: mpsc::Sender<(BoxedAny, oneshot::Sender<Result<BoxedAny, BoxedAny>>)>,
-    id: E::Id,
-    _e: PhantomData<E>,
-    _t: PhantomData<T>,
-}
+    /// Reply type.
+    type Reply: Debug + Send + 'static;
 
-impl<E, T> EntityRef<E, T>
-where
-    E: EventSourced + 'static,
-{
-    /// The ID of the represented [EventSourced] entity.
-    pub fn id(&self) -> &E::Id {
-        &self.id
-    }
+    /// Error type for rejected (a.k.a. invalid) commands.
```

Comment:
Are all those commands really "invalid", isn't more like "currently not applicable".

Reply:
Like `InvalidArgumentExceptiion` ;-)